engine/flink/management/src/main/scala/pl/touk/nussknacker/engine/management/FlinkRestManager.scala

```diff
@@ -119,21 +120,39 @@ class FlinkRestManager(config: FlinkConfig, modelData: ModelData, sender: HttpSe
         .filter(status => !status.state.isGloballyTerminalState || status.state == JobStatus.FINISHED)
 
       runningOrFinished match {
-        case Nil => None
+        case Nil => Future.successful(None)
         case duplicates if duplicates.count(_.state == JobStatus.RUNNING) > 1 =>
-          Some(ProcessState(DeploymentId(duplicates.head.jid), RunningState.Error, "INCONSISTENT", duplicates.head.`start-time`,
-            Some(s"Expected one job, instead: ${runningOrFinished.map(job => s"${job.jid} - ${job.state.name()}").mkString(", ")}")))
+          Future.successful(Some(ProcessState(DeploymentId(duplicates.head.jid), RunningState.Error, "INCONSISTENT", duplicates.head.`start-time`, None,
+            Some(s"Expected one job, instead: ${runningOrFinished.map(job => s"${job.jid} - ${job.state.name()}").mkString(", ")}"))))
         case one::_ =>
           val runningState = one.state match {
             case JobStatus.RUNNING => RunningState.Running
             case JobStatus.FINISHED => RunningState.Finished
             case _ => RunningState.Error
           }
-          Some(ProcessState(DeploymentId(one.jid), runningState, one.state.toString, one.`start-time`))
+          checkVersion(one.jid, name).map { version =>
+            Some(ProcessState(DeploymentId(one.jid), runningState, one.state.toString, one.`start-time`, version))
+          }
       }
     }
   }
 
+  //TODO: cache by jobId?
+  private def checkVersion(jobId: String, name: ProcessName): Future[Option[ProcessVersion]] = {
+    sender.send {
+      (flinkUrl / "jobs" / jobId / "config").GET OK utils.asJson[JobConfig]
+    }.map { config =>
+      val userConfig = config.`execution-config`.`user-config`
+      for {
+        version <- userConfig.get("versionId").flatMap(_.string).map(_.toLong)
```

Comment:
Why you flatMap on this? Does all this things are normal business case or rather they are kind of really unexpected behaviour? If the second one, maybe will be better to throw exception? Looking on this I think about how many strange things could happen (missing versionId, versionId which is not string, versionId which is not string holding long...) and wondering if developer could see here only real business scenarios like: there is correct process config in user config or there isn't.

Reply:
good catch :) it's mainly for backward compatibility of deployed jobs, I added TODO and explanation